src/miniscript/musig_key.rs

```diff
@@ -0,0 +1,166 @@
+//! Support for multi-signature keys
+use core::fmt;
+use core::str::FromStr;
+
+use crate::expression::{FromTree, Tree};
+use crate::prelude::*;
+use crate::{Error, MiniscriptKey};
+
+#[derive(Clone, PartialEq, PartialOrd, Ord, Eq, Hash)]
+/// Enum for representing musig keys in miniscript
+pub enum KeyExpr<Pk: MiniscriptKey> {
+    /// Single-key (e.g pk(a), here 'a' is a single key)
+    SingleKey(Pk),
+
+    /// Collection of keys in used for musig-signature
+    MuSig(Vec<KeyExpr<Pk>>),
+}
+
+impl<Pk: MiniscriptKey + FromStr> FromTree for KeyExpr<Pk> {
+    fn from_tree(tree: &Tree) -> Result<KeyExpr<Pk>, Error> {
+        if tree.name == "musig" {
+            let mut key_expr_vec = vec![];
+            for sub_tree in tree.args.iter() {
+                let temp_res = KeyExpr::<Pk>::from_tree(sub_tree)?;
+                key_expr_vec.push(temp_res);
+            }
+            Ok(KeyExpr::MuSig(key_expr_vec))
+        } else {
+            let single_key = Pk::from_str(tree.name).map_err(|_| Error::SingleKeyParseError)?;
+            Ok(KeyExpr::SingleKey(single_key))
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey + FromStr> FromStr for KeyExpr<Pk> {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (key_tree, _) = Tree::from_slice(s).unwrap();
```

Comment:
I think we should propagate the error rather than panicking here. From what little I understand, we only use `unwrap` in cases we're sure errors shouldn't occur. In this case, `s` might be a malformed string and should be handled by the calling function itself.

Reply:
Yes, this seems the right way of handling the error in this case.
Done.